File: app/csharp-ms/ManagementAccounting/ManagementAccounting.Infrastructure/Persistence/FinancialAnalysisCacheRepository.cs
using Dapper;
using ManagementAccounting.Application.Ports.Out;
using ManagementAccounting.Domain.Entities;
using Microsoft.Extensions.Configuration;
using Npgsql;

namespace ManagementAccounting.Infrastructure.Persistence;

/// <summary>
/// 財務分析キャッシュリポジトリ（Dapper 実装）
/// </summary>
public class FinancialAnalysisCacheRepository : IFinancialAnalysisCacheRepository
{
    private readonly string _connectionString;

    public FinancialAnalysisCacheRepository(IConfiguration configuration)
    {
        _connectionString = configuration.GetConnectionString("ManagementAccounting")!;
    }

    public async Task<FinancialAnalysisCache> SaveAsync(FinancialAnalysisCache cache)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        // UPSERT: 既存レコードがあれば更新、なければ挿入
        var sql = @"
            INSERT INTO financial_analysis_cache
                (fiscal_year, sales, operating_profit, operating_profit_margin,
                 total_asset_turnover, equity_ratio, calculated_at)
            VALUES
                (@FiscalYear, @Sales, @OperatingProfit, @OperatingProfitMargin,
                 @TotalAssetTurnover, @EquityRatio, @CalculatedAt)
            ON CONFLICT (fiscal_year)
            DO UPDATE SET
                sales = EXCLUDED.sales,
                operating_profit = EXCLUDED.operating_profit,
                operating_profit_margin = EXCLUDED.operating_profit_margin,
                total_asset_turnover = EXCLUDED.total_asset_turnover,
                equity_ratio = EXCLUDED.equity_ratio,
                calculated_at = EXCLUDED.calculated_at
            RETURNING id";

        var id = await connection.ExecuteScalarAsync<int>(sql, cache);
        cache.Id = id;
        return cache;
    }

    public async Task<FinancialAnalysisCache?> FindByFiscalYearAsync(int fiscalYear)
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = @"
            SELECT id, fiscal_year AS FiscalYear, sales, operating_profit AS OperatingProfit,
                   operating_profit_margin AS OperatingProfitMargin,
                   total_asset_turnover AS TotalAssetTurnover,
                   equity_ratio AS EquityRatio, calculated_at AS CalculatedAt
            FROM financial_analysis_cache
            WHERE fiscal_year = @FiscalYear";

        return await connection.QuerySingleOrDefaultAsync<FinancialAnalysisCache>(
            sql, new { FiscalYear = fiscalYear });
    }

    public async Task<List<FinancialAnalysisCache>> FindAllAsync()
    {
        await using var connection = new NpgsqlConnection(_connectionString);

        var sql = @"
            SELECT id, fiscal_year AS FiscalYear, sales, operating_profit AS OperatingProfit,
                   operating_profit_margin AS OperatingProfitMargin,
                   total_asset_turnover AS TotalAssetTurnover,
                   equity_ratio AS EquityRatio, calculated_at AS CalculatedAt
            FROM financial_analysis_cache
            ORDER BY fiscal_year";

        var results = await connection.QueryAsync<FinancialAnalysisCache>(sql);
        return results.ToList();
    }
}

File: app/fsharp-ms/src/ManagementAccounting.Infrastructure/Persistence/Repositories/FinancialAnalysisCacheRepository.fs
namespace ManagementAccounting.Infrastructure.Persistence.Repositories

open System
open System.Data
open System.Text.Json
open System.Threading.Tasks
open Npgsql
open Dapper
open ManagementAccounting.Domain.Models
open ManagementAccounting.Application.Ports.Out

/// <summary>
/// DB 行モデル（financial_analysis_cache テーブル）
/// </summary>
[<CLIMutable>]
type FinancialAnalysisCacheRow = {
    id: int
    fiscal_year: int
    data_json: string
    ratios_json: string
    cached_at: DateTime
}

/// <summary>
/// 財務分析キャッシュリポジトリ実装（Dapper）
/// </summary>
type FinancialAnalysisCacheRepository(connectionString: string) =

    let createConnection () : IDbConnection =
        new NpgsqlConnection(connectionString)

    let jsonOptions = JsonSerializerOptions(PropertyNamingPolicy = JsonNamingPolicy.CamelCase)

    let toCache (row: FinancialAnalysisCacheRow) : FinancialAnalysisCache =
        let data = JsonSerializer.Deserialize<FinancialData>(row.data_json, jsonOptions)
        let ratios = JsonSerializer.Deserialize<FinancialRatios>(row.ratios_json, jsonOptions)
        {
            Id = Some row.id
            FiscalYear = row.fiscal_year
            Data = data
            Ratios = ratios
            CachedAt = row.cached_at
        }

    interface IFinancialAnalysisCacheRepository with
        member _.SaveAsync(cache: FinancialAnalysisCache) =
            task {
                use connection = createConnection()
                connection.Open()

                let dataJson = JsonSerializer.Serialize(cache.Data, jsonOptions)
                let ratiosJson = JsonSerializer.Serialize(cache.Ratios, jsonOptions)

                // UPSERT: 既存のキャッシュがあれば更新、なければ挿入
                let sql = """
                    INSERT INTO financial_analysis_cache (fiscal_year, data_json, ratios_json, cached_at)
                    VALUES (@FiscalYear, @DataJson, @RatiosJson, @CachedAt)
                    ON CONFLICT (fiscal_year)
                    DO UPDATE SET
                        data_json = @DataJson,
                        ratios_json = @RatiosJson,
                        cached_at = @CachedAt
                    RETURNING id
                """

                let parameters = {|
                    FiscalYear = cache.FiscalYear
                    DataJson = dataJson
                    RatiosJson = ratiosJson
                    CachedAt = cache.CachedAt
                |}

                let! id = connection.ExecuteScalarAsync<int>(sql, parameters)
                return { cache with Id = Some id }
            }

        member _.GetByFiscalYearAsync(fiscalYear: int) =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = "SELECT * FROM financial_analysis_cache WHERE fiscal_year = @FiscalYear"
                let! results = connection.QueryAsync<FinancialAnalysisCacheRow>(sql, {| FiscalYear = fiscalYear |})

                return
                    results
                    |> Seq.tryHead
                    |> Option.map toCache
            }

        member _.DeleteByFiscalYearAsync(fiscalYear: int) =
            task {
                use connection = createConnection()
                connection.Open()

                let sql = "DELETE FROM financial_analysis_cache WHERE fiscal_year = @FiscalYear"
                let! rowsAffected = connection.ExecuteAsync(sql, {| FiscalYear = fiscalYear |})
                return rowsAffected > 0
            }
